aips/aip-x.md

```diff
@@ -0,0 +1,192 @@
+---
+aip: (this is determined by the AIP Manager, leave it empty when drafting)
+title: Aptos Intent Framework
+author: runtian-zhou
+discussions-to (*optional): <a url pointing to the official discussion thread>
+Status: Draft
+last-call-end-date (*optional): <mm/dd/yyyy the last date to leave feedbacks and reviews>
+type: Application, Framework
+created: 09/27/2024
+updated (*optional): <mm/dd/yyyy>
+requires (*optional): <AIP number(s)>
+---
+
+# AIP-X - Aptos Intent Framework
+
+## Summary
+
+Designing a generic mechanism for declaring intents for trading between on chain resources on Aptos. This would allow users to declare which operations they would like to perform upfront and let 3rd party builders to fill in the details on how this transactions should be executed.
+
+### Out of scope
+
+We will only be covering the intent declaration and execution logic. We are not going to touch the solving logic on how intents should be fulfilled and we will leave it for builders on Aptos to figure out.
+
+The current design of the intent framework facilitates on chain declaration of intents to simplify the problem. Ideally the intent should be declarable offchain as well using signed message but this is beyond the scope of the current AIP. We will provide a path on how we can enable features like this in the future.
+
+The intent system designed here also focus on the Aptos only setup and we are not looking into cross chain solutions so far.
+
+## High-level Overview
+
+We see intent as a programmable lock for on chain assets. Meaning:
+- One party can declare:
+  - Type of Asset they would like to trade in
+  - Condition when this asset could be released
+- The framework provides mechanism to release the lock when such condition is met.
+
+Logic here can then be translated into a Move struct that looks like the following:
+```
+struct TradeIntent<Source, phantom Target, Args> has key {
+    offered_resource: Source,
+    argument: Args,
+    // Fn(Target, Args) -> ()
+    consumption_function: FunctionInfo,
+    self_delete_ref: DeleteRef,
+    expiry_time: u64,
+}
+```
+
+The `TradeIntent` is parameterized by three type parameters:
+1. `Source`: Value that intent issuer is willing to give away
+2. `Target`: the type of value that the intent issuer is expecting.
+3. `Argument`: Intent user will use `Argument` as auxillary info to determine whether the unlock condition is met.
```

Comment:
Do you intend for `argument` to be a proof (zk-proof or storage proof), which can be verified on-chain?

Reply:
In the current design, there's no proof involved: user specifies the post condition the transaction will need to meet, and aptos framework will assert this transaction can only be committed if post condition has been met. `Argument` is just some auxillary data to help the framework determine whether the transaction can be committed. It would be easier if you look at the move code directly.